File: bridge/src/main/java/ru/hollowhorizon/hollowengine/api/RegistryProvider.java
package ru.hollowhorizon.hollowengine.api;

import net.minecraft.core.Registry;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.Supplier;

public interface RegistryProvider<T> {
    RegistryHolder<T> register(
            @NotNull ResourceLocation location,
            @Nullable Registry<T> registry,
            @Nullable AutoModelType model,
            @NotNull Supplier<T> generator,
            @NotNull Class<T> type
    );
}

File: runtime/src/main/java/ru/hollowhorizon/hollowengine/common/registry/RegistryProvider.kt
package ru.hollowhorizon.hollowengine.common.registry

import net.minecraft.core.Registry
import net.minecraft.resources.ResourceLocation
import ru.hollowhorizon.hollowengine.api.AutoModelType
import ru.hollowhorizon.hollowengine.api.RegistryHolder
import ru.hollowhorizon.hollowengine.api.RegistryProvider
import java.util.function.Supplier

object CommonRegistryProvider : RegistryProvider<Any> {
    private lateinit var registry: RegistryProvider<Any>

    fun init(registry: RegistryProvider<Any>) {
        this.registry = registry
    }

    override fun register(
        location: ResourceLocation,
        registry: Registry<Any>?,
        model: AutoModelType?,
        generator: Supplier<Any>,
        type: Class<Any>,
    ): RegistryHolder<Any> {
        return this.registry.register(location, registry, model, generator, type)
    }
}